# Problem of the week #7 due 11/10/21

The integers from 1 to 2021 are written in order, in one long row, with no spaces between them.
`1234567891011…20202021`
What is the middle digit?

## Short Answer

The middle digit in the string of all integers from 1 to 2021 with no spaces between them is **9**.

## Method

I am going to be solving this problem with pure math, but I will be checking my answers with code.

This paper is divided into 2 parts:

    1. Calculating the number of digits, and then finding the index of the middle digit
    2. Finding out what number is at the index of that middle digit
    
We solve part 1 by adding up the total number of digits contributed to the total by each 'type' of integer. We define a type of integer as having the same number of digits as all other integers of the same type. Thus there are 4 types of integers in this problem, integers with 1, 2, 3, and 4 digits.

We solve part 2 by simplifying the problem, determining how to find any nth digit in the simplified version, and finally using this to find the answer to the problem.

## Getting Started

I am going to start by defining a python function that will return the combined string of numbers between any beginning and end number, inclusive, that we specify. I am defining this function up-front so that we can use it throughout the paper. It will return both the actual string of numbers, and the number of numbers in the string, in a dictionary data structure.

In [808]:
import math

In [809]:
def number_string(beginning, end):
    accumulator = ""
    number_of_numbers = 0
    for i in range(beginning, end+1):
        accumulator+=str(i)
        number_of_numbers+=1
    return {
        "string": accumulator,
        "number_of_numbers": number_of_numbers
    }

Let's go ahead and check this code by having it print the numbers 1 through 5 for us.

For the string that it returns, we expect to see `12345`, and for the number of numbers that it returns, we expect the number 5.

In [810]:
number_string(1,5)

{'string': '12345', 'number_of_numbers': 5}

We can use the python `len()` function to get the length of the string. When we take the length of the string returned by our `number_string` function, it tells us the total number of digits. For 1 digit numbers only, this will be the same as the number of numbers, but once we get into higher digit numbers, this will be the key function. Lets see how we get the number of digits for our string of 1 through 5:

In [811]:
len(number_string(1,5)['string'])

5

## How many digits are in our full number string?

This is all integers 1 through 2021 

We can divide the integers 1 through 2021 into 4 different categories based on their number of digits. We can refer to each category as a 'type' of integer. Each integer will have the same number of digits as all others of its same type.

The number of digits in the full string of integers 1 through 2021, without spaces, will be the sum of the number of digits of the strings made up by each type of integer, because each integer 1 to 2021 belongs to 1 type and only one type. So in this will, each integer will still be counted once and only one. 

It makes sense to solve this problem in this way because all we have to do is figure out how many integers belong to each type, and then multiply that by the number of digits in each integer of that type, to calculate the total digits in the string of integers of that type, without spaces.

### How Many Digits do the One Digit Numbers Make Up?

Well, the only 1 digit integers are 1 through 9. The number of these numbers is not, however 9-1, because the string is inclusive. In order to get the actual number of numbers in the string, we have to either subtract 1 from the beginning number, or add 1 to the highest number. I will be using the former method throughout this paper.

Using this formula, we get:

In [812]:
9-(1-1)

9

For one digit numbers, the total number of digits will be equal to the number of numbers, because each number contributes exactly 1 digit.

Lets now double check this with our python function we defined earlier

In [813]:
one_digit_numbers = number_string(1,9)
print("Number String:", one_digit_numbers['string'])
print("Number of Numbers:", one_digit_numbers['number_of_numbers'])
print("Number of Digits:", len(one_digit_numbers['string']))

Number String: 123456789
Number of Numbers: 9
Number of Digits: 9


### How Many Digits do the Two Digit Numbers Make Up?

All we have to do, is follow the same formula that we did for the one digit numbers above. That is, the last 2 digit number minus one less than the first 2 number.

In [814]:
99-(10-1)

90

In order to get the number of digits of all of these 2 digit numbers combined, we just multiply the number of numbers by the number of digits in each, which is of course 2 

In [815]:
2 * 90

180

Double checking this with our python function:

In [816]:
two_digit_numbers = number_string(10,99)
print("Number String:", two_digit_numbers['string'], '\n')
print("Number of Numbers:", two_digit_numbers['number_of_numbers'])
print("Number of Digits:", len(two_digit_numbers['string']))

Number String: 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899 

Number of Numbers: 90
Number of Digits: 180


We can see here that so far our math is matching our code output perfectly. We can reality check our code by looking at the number string, and seeing that it does indeed start with 10 and end with 99.

### How Many Digits do the Three Digit Numbers Make Up?

The first 3 digit number is 100, and the last is 999. Plugging this into the formulas discussed above, we can calculate:

The number of numbers:

In [817]:
999 - (100-1)

900

And thus the number of digits (each number contributes 3 digits):

In [818]:
900 * 3

2700

Now to double check this with our python function. In this case, though, I am not going to print out the number string, because it takes up too much space and looks messy.

In [819]:
three_digit_numbers = number_string(100,999)
print("Number of Numbers:", three_digit_numbers['number_of_numbers'])
print("Number of Digits:", len(three_digit_numbers['string']))

Number of Numbers: 900
Number of Digits: 2700


### How Many Digits do the Four Digit Numbers Make Up?

The first 4 digit number is 1000, and the last is 2021.

The number of numbers:

In [820]:
2021 - (1000-1)

1022

The number of digits (each number contributes 4 digits):

In [821]:
1022 * 4

4088

Double checking this with our python function, and once again not printing the full number string:

In [822]:
four_digit_numbers = number_string(1000,2021)
print("Number of Numbers:", four_digit_numbers['number_of_numbers'])
print("Number of Digits:", len(four_digit_numbers['string']))

Number of Numbers: 1022
Number of Digits: 4088


### Combining all of our Digits

In [823]:
9 + 180 + 2700 + 4088

6977

We can also use all of the variables we have defined thus far in the following formula:

In [824]:
one_digit_numbers['number_of_numbers'] * 1 + \
two_digit_numbers['number_of_numbers'] * 2 + \
three_digit_numbers['number_of_numbers'] * 3 + \
four_digit_numbers['number_of_numbers'] * 4

6977

Or we can just use our function to generate the full string from 1 to 2021, and check the length of that

In [825]:
len(number_string(1,2021)['string'])

6977

In all 3 cases, we find that the number of digits in the string from 1 to 2021 is **6977**

## Finding the Middle Digit

The first thing that can be noticed about the number of digits is that it is an odd number.

This helps us to find the middle number. Consider the string of integers 1 through 5 (`12345`), half of 5 is 2.5, but the middle integer is clearly 3, which is the third digit. So to find the middle digit, we divide the total by 2, and round up.

We can use the `math.ceil` function in python to round up. Thus, the middle digit of our string of numbers from 1 to 2021 is:

In [826]:
math.ceil(6977 / 2)

3489

### What is the 3489th digit in our string?

Unfortunately, unlike in our simple example of `12345` the nth digit is not guarunteed to be n once we get into higher numbers.

To start on finding the 3489th digit, I am going to simplify our problem by removing all of the non-four digit numbers. How to ignore the 1, 2, and 3 digit numbers? All we have to do is subtract the total number of digits contributed by the 1, 2, and 3 digit numbers from 3489. This will tell us the number of digits into the string of only 4 digit numbers that we have to look.

We know that we can do this because for any string $$a = a_0a_1a_2a_3...a_r$$ removing the first n elements will yield a new string b such that $$b = b_0b_1b_2b_3...b_s =a_na_{n+1}a_{n+2}...a_r$$ and $$a_i=b_{i-n}$$

So how many digits do 1, 2, and 3 digit numbers contribute?
We can look up above at the sections corresponding to those, and get:

In [827]:
9 + 180 + 2700

2889

We can double check this with our code by generating the string of integers from the first one digit number to the last 3 digit number, and calculate the length of that.

In [828]:
len(number_string(1,999)['string'])

2889

Now we can subtract this number from 3489, to get the digit that we want within the string made up of only 4 digit numbers 1000 to 2021, inclusive.

In [829]:
3489 - 2889

600

Note: if this number had been negative, then we would have known that the middle digit of the total string did not lie within the portion of the string made up of 4 digit numbers, and we would have had to find a different method to proceed with.

### What is the nth digit of any string of numbers?

Lets start with a simpler problem, and ask, what is the 6th digit of this series? Well, to start, we don't need the whole series in order to figure this out, so we can just consider the string of 1000 to 1003. I am going to use python to display this string for us:

In [830]:
number_string(1000, 1003)['string']

'1000100110021003'

We count the first 1 as the 1st digit, then the following 0 as the second, then the following zero as the 3rd, then finally the last zero in 1000 as the 4th digit. Then starts the integer that we are interested in, because the 6th digit will lie somewhere in the year 1001.

We can notice here that we can deduce the year that we care about by using the digit divided by 4. In python, we can use the `math.floor` function to round this down to the nearest integer. In the case of the sixth digit, we use

In [831]:
math.floor(6 / 4)

1

to find that it is the first year after 1000, that our digit lies in.

To find the digit within that year that we care about, we can use modular arithmetic. The modulo function (in python it is `%`), will give us the remainder after division. Thus we can use n % 4 to find the remainder after dividing by 4. This will tell us which digit we care about within the year that we care about.

In the example of the sixth digit, we use

In [832]:
6 % 4

2

This is tell us that it is the second digit in the year 1001, that is the sixth digit in the string of years 1000 to 1003. Thus, the sixth digit is 0.


**Important Note:** the caveat to using this method is that `n % 4` will never equal 4. So how do we get the fourth digit? Well, if we try to look for the fourth digit, we will do

In [833]:
math.floor(4 / 4)

1

which should tell us it is the year 1001 that we want, but we can see that

In [834]:
4 % 4

0

We have to interpret this zero as telling us to look at the last digit of the year previous to the one the former equation told us to care about. In this case, the last digit of the year 1000. Thus the fourth digit is 0, which is easily verifiable manually.

#### Another Example

Lets try this again to find the 13th digit.
The year that we care about is:

In [835]:
math.floor(13 / 4)

3

years after 1000, which is 1003.

The digit within that year is:

In [836]:
13 % 4

1

Thus, the 13th digit is the first digit in 1003, which is 1. We can check this by manually counting out 13 in `1000100110021003`

#### Checking indexes in Python

An important note about python is that it will start counting strings starting at 0. For example, in the string `abcde`, python interprets the first index as being the second character in the string. As we can see here:

In [837]:
"abcde"[1]

'b'

 Thus in order to get python to tell us what we think of as the first character in the string, we always have to subtract 1. Here is the expected output, achieved by subtracting 1:

In [838]:
"abcde"[1-1]

'a'

Thus, to have python check the 13th character in the string of numbers from 1000 to 1003, we have to ask it for the (13 - 1)th character.

In [839]:
number_string(1000, 1003)['string'][13-1]

'1'

### What is the 600th digit of the string made up of numbers 1000 to 2021?

Now to answer the real question, all we have to do is use the formulas derived above on `600`

The year we care about is

In [840]:
math.floor(600 / 4)

150

years after 1000, which is 1150.

The digit we care about within that year is:

In [841]:
600 % 4

0

In accordance with our important note earlier, this 0 tells us that we actually care about the last digit of the year preceeding 1150. Now, by hand we can see that the last digit of the year 1149 is **9**, which is our final answer.

### Double Checking with Python

We can double check this using python in a number of ways, but lets start by calculating the 600th digit of the string of integers from 1000 to 2021:

In [842]:
number_string(1000,2021)["string"][600-1]

'9'

Another way of using python to check our solution is to generate the whole string of 1 to 2021, and then pull the middle digit from that:

In [843]:
full_string = number_string(1,2021)['string']
number_of_digits = len(full_string)
middle_digit_index = math.ceil(number_of_digits/2)
middle_digit = full_string[middle_digit_index-1]

print(middle_digit)

9


Alternatively, we can totally ignore readability and collapse this code into a single line. Thus,

## Here is one line of code that can solve the problem

*given the python function defined at the start

In [844]:
number_string(1,2021)['string'][math.ceil(len(number_string(1,2021)['string'])/2)-1]

'9'